2996_reflection/reflection.md

```diff
@@ -2283,6 +2284,86 @@ Lastly, we clarify that during the definition of an _injected declaration_, the
+  }
+};
+
+static_assert(TCls<void>::sfn());
+static_assert(is_complete_type(^^S));
+```
+
+The above example observes the effects of the failed substitution of `#1` by way of the completeness of `S`. Such tricks can be used to observe implementation details, like the order in which overloads are checked, that may be unportable (and which implementations might desire to change over time).
+
+Our proposed solution, specified in [expr.const]/23.2, is to make it ill-formed to produce an injected declaration _from_ a manifestly constant-evaluated expression within an instantiation to _outside of_ that instantiation, or visa versa. Because that expression in the example above (`define_aggregate(^^S, {})`) is within the instantiation of the requires clause of `TCls<void>::sfn`, and the target scope of the injected declaration is outside of that same instantiaton, the example becomes ill-formed (diagnostic required). Note that this does not prevent writing `consteval` function templates that wrap `define_aggregate`:
+
+```cpp
+template <std::meta::info R> consteval bool tfn() {
+  define_aggregate(R, {});
+  return true;
+}
+
+struct S;
+constexpr bool b = tfn<^^S>();
+  // OK, both manifestly constant-evaluated expression tfn<^^S>() and target scope of
+  // injected declaration for 'S' are in the global namespace
+```
+
+Nor does this rule prevent a class template from producing a declaration whose target scope is the same specialization.
+
+```cpp
+template <typename> struct TCls1 {
+  struct Incomplete;
+
+  consteval {
+    define_aggregate(^^Incomplete, {});
+      // OK, Incomplete is in the same instantiation as the define_aggregate call
+  }
+
+  static constexpr bool b = false;
+};
+
+template <typename T> struct TCls2 {
+  static consteval bool sfn()  // #1
+      requires (TCls1<T>::b) {
+    return false;  // never selected
+  }
+
+  static consteval bool sfn()  // #2
+      requires (true) {
+    return true;   // always selected
+  }
+};
+
+static_assert(TCls<void>::sfn());
+```
+
+Athough the instantiation of `TCls1<void>` in the requires-clause of `#1` causes an injected declaration to be produced, it is not discernibly a side-effect of the failed substitution: Observing the side effect will first require one to write (some moral  equivalent of) `TCLs1<void>::Incomplete`, the act of which would otherwise itself trigger the same side-effect.
+
+Although this rule constrains the manner with which `define_aggregate` can be used, we are not aware of any motivating use cases for P2996 that are harmed. Worth mentioning, however: the rule has more dire implications for other code injection papers being considered by WG21, most notably P3294 ("_Code Injection With Token Sequences_"). With this rule as it is, it becomes impossible for e.g., the instantiation of a class template specialization `TCls<Foo>` to produce an injected declaration of `std::formatter<TCls<Foo>>` (since the target scope would be the global namespace).
```

Comment:
Use [@P3294R2] instead of just P3294, which will add the link.

Reply:
Yep, thanks.